src/wasm/wasm-type.cpp

```diff
@@ -1087,6 +955,233 @@ std::ostream& operator<<(std::ostream& os, HeapTypeInfo info) {
   WASM_UNREACHABLE("unexpected kind");
 }
 
+namespace {
+
+bool TypeComparator::cmp(Type a, Type b) {
+  if (a == b) {
+    return false;
+  }
+  if (a.isBasic() && b.isBasic()) {
+    return a.getBasic() < b.getBasic();
+  }
+  if (a.isBasic()) {
+    return true;
+  }
+  if (b.isBasic()) {
+    return false;
+  }
+  return cmp(*getTypeInfo(a), *getTypeInfo(b));
+}
+
+bool TypeComparator::cmp(HeapType a, HeapType b) {
+  if (a == b) {
+    return false;
+  }
+  if (cache.count({a, b})) {
+    return true;
+  }
+  if (a.isBasic() && b.isBasic()) {
+    return a.getBasic() < b.getBasic();
+  }
+  if (a.isBasic()) {
+    return true;
+  }
+  if (b.isBasic()) {
+    return false;
+  }
+  cache.insert({a, b});
+  return cmp(*getHeapTypeInfo(a), *getHeapTypeInfo(b));
```

Comment:
I'm not sure I follow what `cache` does. Trying to understand from the code, on the one hand if `(x,y)` are in the cache then we return true. However right here we insert `(x,y)` into the cache regardless of the result, that is, we may end up returning false. So it is not purely caching the function result - is that right?

Reply:
Oh, I think I see, it's not a cache of the results, but a recursion guard. Perhaps `cache => seen`?

Also a comment on the return when something is found in `cache/seen` could explain why the result does not matter in that case. It's not immediately obvious to me why the result is deterministic in that case.

Reply:
Yeah, sorry, this code could use more comments in general. Will fix it up.